### PR TITLE
MODKBEKBJ-479 - BE | Package | Export Titles

### PR DESCRIPTION
**Pull request by natali-zaitseva**
## Purpose
implementation story for https://issues.folio.org/browse/MODKBEKBJ-479 supports to export titles with default sort by name ascending order

## Approach
* implemented endpoint to export titles tied to package
* added tests
* created raml for export
* updated ModuleDescriptor file

### CODE REVIEW

File: src/main/java/org/folio/service/uc/export/ExportServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    // mapping of columns by position
+    CustomBeanToCSVMappingStrategy<TitleExportModel> mappingStrategy = new CustomBeanToCSVMappingStrategy<>();
+    mappingStrategy.setType(TitleExportModel.class);
+
+    StatefulBeanToCsv<TitleExportModel> csvToBean = new StatefulBeanToCsvBuilder<TitleExportModel>(writer)
+      .withMappingStrategy(mappingStrategy)
+      .withQuotechar(NO_QUOTE_CHARACTER)
+      .withEscapechar(NO_ESCAPE_CHARACTER)
+      .withLineEnd(DEFAULT_LINE_END)
+      .withSeparator('|')
```
Comment (psmagin) [inline src/main/java/org/folio/service/uc/export/ExportServiceImpl.java:55]: Why we use '|' this char as a separator? I think it is not a commonly used separator for CSV files.

File: src/main/java/org/folio/rest/converter/costperuse/export/PackageTitlesCollectionConverter.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import java.util.List;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import org.folio.rest.jaxrs.model.ResourceCostPerUseCollection;
+import org.folio.service.uc.export.TitleExportModel;
+
+@Component
+public class PackageTitlesCollectionConverter {
```
[the remark this answers is not in the document]
Comment (natali-zaitseva) [reply src/main/java/org/folio/rest/converter/costperuse/export/PackageTitlesCollectionConverter.java]: renamed

File: src/main/java/org/folio/service/uc/export/ExportServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    LOG.info("Mapping " + entities.size() + " entities to SCV");
+    StringWriter writer = new StringWriter();
+
+    // mapping of columns by position
+    CustomBeanToCSVMappingStrategy<TitleExportModel> mappingStrategy = new CustomBeanToCSVMappingStrategy<>();
+    mappingStrategy.setType(TitleExportModel.class);
+
+    StatefulBeanToCsv<TitleExportModel> csvToBean = new StatefulBeanToCsvBuilder<TitleExportModel>(writer)
+      .withMappingStrategy(mappingStrategy)
+      .withQuotechar(NO_QUOTE_CHARACTER)
```
[the remark this answers is not in the document]
Comment (natali-zaitseva) [reply src/main/java/org/folio/service/uc/export/ExportServiceImpl.java]: agree with you, fixed

File: src/main/java/org/folio/service/uc/export/TitleExportModel.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  @CsvBindByPosition(position = 1)
+  private final String type;
+
+  @CsvBindByName(column = "Usage")
+  @CsvBindByPosition(position = 2)
+  private final int usage;
+
+  @CsvBindByName(column = "Cost")
+  @CsvBindByPosition(position = 3)
+  private final double cost;
```
[the remark this answers is not in the document]
Comment (natali-zaitseva) [reply src/main/java/org/folio/service/uc/export/TitleExportModel.java:24]: added, thanks, almost forgot about this field

File: src/main/resources/liquibase/tenant/scripts/v3.5.0/alter-holdings-status-table.xml [hunk 5]
```diff
@@ -74,16 +74,16 @@
...
+<!--  <changeSet id="MODKBEKBJ-444@@enable-holdings-status-audit-trigger" author="dmtkachenko">-->
+<!--    <preConditions onFail="MARK_RAN">-->
+<!--      <sqlCheck expectedResult="true">-->
+<!--        SELECT ${database.defaultSchemaName}.trigger_exists('holdings_status_audit');-->
+<!--      </sqlCheck>-->
+<!--    </preConditions>-->
+<!--    <sql>-->
+<!--      ALTER TABLE ${database.defaultSchemaName}.holdings_status ENABLE TRIGGER holdings_status_audit;-->
+<!--    </sql>-->
+<!--  </changeSet>-->
```
[the remark this answers is not in the document]
Comment (natali-zaitseva) [reply src/main/resources/liquibase/tenant/scripts/v3.5.0/alter-holdings-status-table.xml]: done, merged master to "export_poc" branch

File: src/main/java/org/folio/service/uc/export/ExportServiceImpl.java (hunk 1 above)
Comment (natali-zaitseva) [reply src/main/java/org/folio/service/uc/export/ExportServiceImpl.java:53]: we use "|" character to align with the same separator as EBSCO uses, just not to confuse users. We can change the separator to any other character we agreed.

File: ramls/types/publicationType.json [hunk 6]
```diff
@@ -19,8 +20,10 @@
         "Streaming Audio",
         "Streaming Video",
         "Thesis & Dissertation",
+        "Thesis/Dissertation",
         "Website",
+        "Web site",
```
[the remark this answers is not in the document]
Comment (natali-zaitseva) [reply ramls/types/publicationType.json:25]: for some reason, APIGEE service returns Publication types in this format, currently, I have faced several of them and "Web site" is one of them. as well as "Audio book" , "Thesis/Dissertation"

[the remark this answers is not in the document]
Comment (psmagin) [reply ramls/types/publicationType.json:25]: APIGEE doesn't return any publication types. This data is fetched from the holdings table. So this misspelling is on the RMAPI side. I think we should decide how to handle this:
* stay as it is
* make mapping between holdings loading data and data returned from /titles rmapi endpoint

File: src/main/java/org/folio/service/uc/export/TitleExportModel.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
+package org.folio.service.uc.export;
+
+import com.opencsv.bean.CsvBindByName;
+import com.opencsv.bean.CsvBindByPosition;
+import lombok.experimental.SuperBuilder;
+
+@SuperBuilder(toBuilder = true)
```
Comment (dmtkachenko) [inline src/main/java/org/folio/service/uc/export/TitleExportModel.java:7]: What is the reason behind using `SuperBuilder`?
Is this class going to be extended?

File: src/main/java/org/folio/service/uc/export/ExportServiceImpl.java (hunk 1 above)
Comment (psmagin) [reply src/main/java/org/folio/service/uc/export/ExportServiceImpl.java:53]: As I remember in the example that Khalilla showed us this separator was used for data in one column. For example for ISBN column if there several ISBN's for title

File: src/main/java/org/folio/service/uc/export/TitleExportModel.java (hunk 7 above)
Comment (natali-zaitseva) [reply src/main/java/org/folio/service/uc/export/TitleExportModel.java:7]: @dmtkachenko fixed

**Review by dmtkachenko — APPROVED** [review]

File: ramls/types/publicationType.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (natali-zaitseva) [reply ramls/types/publicationType.json:25]: for the current story, I would stay as it is now, but we can create a story for refactoring

**Review by psmagin — APPROVED** [review]

**Pull request merged**